accord-core/src/test/java/accord/impl/basic/InMemoryJournal.java

```diff
@@ -308,29 +325,77 @@ public void purge(CommandStores commandStores)
+                    List<Diff> copy = new ArrayList<>();
+                    int tmp1 = random.nextInt(diffs.size());
+                    int tmp2 = random.nextInt(diffs.size());
+                    int min = Math.min(tmp1, tmp2);
+                    int max = Math.max(tmp1, tmp2);
+                    for (int i = min; i < max; i++)
+                        copy.add(diffs.get(i));
+
+                    if (copy.isEmpty())
+                        continue;
+                    diffs = copy;
+                }
                 InMemoryJournal.Builder builder = reconstruct(diffs, ALL);
+                if (builder.saveStatus() == null)
+                    continue; // Partial compaction, no save status present
+
                 if (builder.saveStatus().status == Truncated || builder.saveStatus().status == Invalidated)
                     continue; // Already truncated
 
-                Command command = builder.construct(store.unsafeGetRedundantBefore());
-                Cleanup cleanup = Cleanup.shouldCleanup(FULL, store.agent(), command, store.unsafeGetRedundantBefore(), store.durableBefore());
-                switch (cleanup)
+                Input input = isPartialCompaction ? PARTIAL : FULL;
+                Cleanup cleanup = shouldCleanup(input,
+                                                store.agent(),
+                                                builder.txnId(),
+                                                builder.executeAt(),
+                                                builder.saveStatus(),
+                                                builder.durability(),
+                                                builder.participants(),
+                                                store.unsafeGetRedundantBefore(),
+                                                store.durableBefore());
+
+                if (builder.maybeCleanup(input, cleanup))
                 {
-                    case NO:
-                        break;
-                    case INVALIDATE:
-                    case TRUNCATE_WITH_OUTCOME:
-                    case TRUNCATE:
-                    case ERASE:
-                        command = Commands.purgeUnsafe(store, command, cleanup);
-                        Invariants.require(command.saveStatus() != SaveStatus.Uninitialised);
-                        Diff diff = toDiff(new CommandUpdate(null, command));
-                        e2.setValue(cleanup == Cleanup.ERASE ? new ErasedList(diff) : new TruncatedList(diff));
-                        break;
-
-                    case EXPUNGE:
-                        e2.setValue(new PurgedList(e2.getValue()));
-                        break;
+                    Invariants.require(cleanup != NO);
+                    if (cleanup == EXPUNGE)
+                        e2.setValue(new PurgedList());
+                    else
+                    {
+                        int flags = CommandChange.flagsForPurge(cleanup.appliesIfNot);
```

Comment:
since we're calling setChanged(SAVE_STATUS, flags) below, I think we can simply use the `eraseKnownFieldsMask` directly?